core/api/src/tasks/telemetry.ts

```diff
@@ -13,14 +16,70 @@ export class Telemetry extends Task {
   async run() {
     if (!config.telemetry.enabled) return;
 
-    const payload = await api.telemetry.build();
+    try {
+      const payload = await api.telemetry.build();
 
-    const response = await fetch(config.telemetry.url, {
-      method: "POST",
-      headers: { "Content-Type": "application/json" },
-      body: JSON.stringify(payload),
-    });
+      const response = await fetch(config.telemetry.url, {
+        method: "POST",
+        headers: { "Content-Type": "application/json" },
+        body: JSON.stringify(payload),
+      });
 
-    return response.status === 200;
+      return response.status === 200;
+    } catch (error) {
+      try {
+        // If there's something wrong with our ability to gather telemetry,
+        // we should try to report that error to the Telemetry service...
+        await fetch(config.telemetry.url, {
+          method: "POST",
+          headers: { "Content-Type": "application/json" },
+          body: JSON.stringify(await this.generateErrorPayload(error)),
+        });
+      } catch (newError) {}
+
+      throw error;
+    }
+  }
+
+  async generateErrorPayload(error) {
+    const clusterName = (await plugin.readSetting("core", "cluster-name"))
+      .value;
+    const customerGuid = (
+      await plugin.readSetting("telemetry", "customer-guid")
+    ).value;
+    const customerLicense = (
+      await plugin.readSetting("telemetry", "customer-license")
+    ).value;
+
+    const errorPayload = {
+      name: clusterName,
+      guid: customerGuid,
+      license: customerLicense,
+      metrics: [
+        {
+          collection: "telemetry",
+          topic: "error",
+          aggregation: "exact",
+          key: `${process.platform}/${os.release()}`,
+          value: this.sanitizeErrorPayload(error),
+        },
+      ],
+    };
+
+    return errorPayload;
+  }
+
+  private sanitizeErrorPayload(error: Error): string {
+    if (!error.stack) return error.toString();
+
+    return error.stack
+      .split(os.EOL)
+      .map((line: string) => {
+        // It's likely the part of the path with 'grouparoo' in it would be safely down in the path structure to not include app-information
+        // There may be many instances of the word "grouparoo" in the file system, so we'll only take the last part of the path
+        const chunks = line.split("grouparoo" + path.sep);
+        return chunks[chunks.length - 1];
+      })
+      .join(os.EOL);
   }
```

Comment:
@bleonard Can you think of a situation where there won't be `grouparoo/` in the path structure indicating that we are safely within `node_modules`?  We should be within `node_modules/@grouparoo/...` for client installs and `grouparoo/grouparoo/core...` for our own development

Reply:
I suppose the core (monorepo) case doesn't _require_ it to be in a directory called `grouparoo` but it's certainly a convention. If they are running the whole monorepo app it could be `/Users/someone/Desktop/testing/core/...`

Reply:
True.  I guess I’m less worried about monorepo users when compared to normal “npm” users.  Have you heard about folks using custom NPM module paths or something like that?